runtime/common/src/types.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright 2024 Moonbeam Foundation.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+use parity_scale_codec::{Decode, Encode, EncodeLike, MaxEncodedLen};
+use scale_info::TypeInfo;
+use sp_std::prelude::*;
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Encode, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(LOWER, UPPER))]
+pub struct BoundedU128<const LOWER: u128, const UPPER: u128>(u128);
+
+impl<const L: u128, const U: u128> BoundedU128<L, U> {
+	pub fn new(value: u128) -> Result<Self, &'static str> {
+		if value < L || value > U {
+			return Err("Value out of bounds");
+		}
+		Ok(Self(value))
+	}
+
+	pub fn new_or_min(value: u128) -> Self {
+		if value < L || value > U {
+			Self(L)
+		} else {
+			Self(value)
+		}
+	}
```

Comment:
I would change this method to:

```suggestion
	pub fn safe_new(value: u128) -> Self {
		if value < L {
			Self(L)
		} else if value > U {
			Self(U)
		} else {
			Self(value)
		}
	}
```

Reply:
We should use the upper bound if `value` is higher than `U`

Reply:
It seems more safe to always use the same bound (lower or upper) if the value is out of bounds, it make the behavior more predictable

Reply:
This method is not used unless you are creating the bounded u128 in the code, (for example creating the default parameter)